// @flatfile/api/dist/serialization/resources/environments/types/Environment.js
"use strict";
/**
 * This file was auto-generated by Fern from our API Definition.
 */
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || function (mod) {
    if (mod && mod.__esModule) return mod;
    var result = {};
    if (mod != null) for (var k in mod) if (k !== "default" && Object.prototype.hasOwnProperty.call(mod, k)) __createBinding(result, mod, k);
    __setModuleDefault(result, mod);
    return result;
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.Environment = void 0;
const core = __importStar(require("../../../../core"));
const EnvironmentId_1 = require("../../commons/types/EnvironmentId");
const AccountId_1 = require("../../commons/types/AccountId");
const GuestAuthenticationEnum_1 = require("./GuestAuthenticationEnum");
exports.Environment = core.serialization.object({
    id: EnvironmentId_1.EnvironmentId,
    accountId: AccountId_1.AccountId,
    name: core.serialization.string(),
    isProd: core.serialization.boolean(),
    guestAuthentication: core.serialization.list(GuestAuthenticationEnum_1.GuestAuthenticationEnum),
    features: core.serialization.record(core.serialization.string(), core.serialization.any()),
    metadata: core.serialization.record(core.serialization.string(), core.serialization.any()),
    translationsPath: core.serialization.string().optional(),
    namespaces: core.serialization.list(core.serialization.string()).optional(),
    languageOverride: core.serialization.string().optional(),
});
