// @flatfile/api/dist/serialization/resources/jobs/types/JobConfig.js
"use strict";
/**
 * This file was auto-generated by Fern from our API Definition.
 */
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || function (mod) {
    if (mod && mod.__esModule) return mod;
    var result = {};
    if (mod != null) for (var k in mod) if (k !== "default" && Object.prototype.hasOwnProperty.call(mod, k)) __createBinding(result, mod, k);
    __setModuleDefault(result, mod);
    return result;
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.JobConfig = void 0;
const core = __importStar(require("../../../../core"));
const JobType_1 = require("./JobType");
const JobSource_1 = require("./JobSource");
const JobDestination_1 = require("./JobDestination");
const JobUpdateConfig_1 = require("./JobUpdateConfig");
const Trigger_1 = require("./Trigger");
const JobStatus_1 = require("./JobStatus");
const FileId_1 = require("../../commons/types/FileId");
const JobMode_1 = require("./JobMode");
const JobSubject_1 = require("./JobSubject");
const EnvironmentId_1 = require("../../commons/types/EnvironmentId");
const JobPartExecution_1 = require("./JobPartExecution");
const JobId_1 = require("../../commons/types/JobId");
exports.JobConfig = core.serialization.object({
    type: JobType_1.JobType,
    operation: core.serialization.string(),
    source: JobSource_1.JobSource,
    destination: JobDestination_1.JobDestination.optional(),
    config: JobUpdateConfig_1.JobUpdateConfig.optional(),
    trigger: Trigger_1.Trigger.optional(),
    status: JobStatus_1.JobStatus.optional(),
    progress: core.serialization.number().optional(),
    fileId: FileId_1.FileId.optional(),
    mode: JobMode_1.JobMode.optional(),
    input: core.serialization.record(core.serialization.string(), core.serialization.any()).optional(),
    subject: JobSubject_1.JobSubject.optional(),
    outcome: core.serialization.record(core.serialization.string(), core.serialization.any()).optional(),
    info: core.serialization.string().optional(),
    managed: core.serialization.boolean().optional(),
    environmentId: EnvironmentId_1.EnvironmentId.optional(),
    part: core.serialization.number().optional(),
    partData: core.serialization.record(core.serialization.string(), core.serialization.any()).optional(),
    partExecution: JobPartExecution_1.JobPartExecution.optional(),
    parentId: JobId_1.JobId.optional(),
    predecessorIds: core.serialization.list(JobId_1.JobId).optional(),
});
