// @flatfile/api/dist/serialization/resources/files/types/File_.js
"use strict";
/**
 * This file was auto-generated by Fern from our API Definition.
 */
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || function (mod) {
    if (mod && mod.__esModule) return mod;
    var result = {};
    if (mod != null) for (var k in mod) if (k !== "default" && Object.prototype.hasOwnProperty.call(mod, k)) __createBinding(result, mod, k);
    __setModuleDefault(result, mod);
    return result;
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.File_ = void 0;
const core = __importStar(require("../../../../core"));
const FileId_1 = require("../../commons/types/FileId");
const ModelFileStatusEnum_1 = require("./ModelFileStatusEnum");
const Mode_1 = require("./Mode");
const SpaceId_1 = require("../../commons/types/SpaceId");
const WorkbookId_1 = require("../../commons/types/WorkbookId");
const SheetId_1 = require("../../commons/types/SheetId");
const Action_1 = require("../../commons/types/Action");
const FileOrigin_1 = require("./FileOrigin");
exports.File_ = core.serialization.object({
    id: FileId_1.FileId,
    name: core.serialization.string(),
    ext: core.serialization.string(),
    mimetype: core.serialization.string(),
    encoding: core.serialization.string(),
    status: ModelFileStatusEnum_1.ModelFileStatusEnum,
    mode: Mode_1.Mode.optional(),
    size: core.serialization.number(),
    bytesReceived: core.serialization.number(),
    createdAt: core.serialization.date(),
    updatedAt: core.serialization.date(),
    expiredAt: core.serialization.date().optional(),
    spaceId: SpaceId_1.SpaceId,
    workbookId: WorkbookId_1.WorkbookId.optional(),
    sheetId: SheetId_1.SheetId.optional(),
    actions: core.serialization.list(Action_1.Action).optional(),
    origin: FileOrigin_1.FileOrigin.optional(),
});
