// @flatfile/api/api/resources/events/types/EventTopic.js
"use strict";
/**
 * This file was auto-generated by Fern from our API Definition.
 */
Object.defineProperty(exports, "__esModule", { value: true });
exports.EventTopic = void 0;
exports.EventTopic = {
    AgentCreated: "agent:created",
    AgentUpdated: "agent:updated",
    AgentDeleted: "agent:deleted",
    SpaceCreated: "space:created",
    SpaceUpdated: "space:updated",
    SpaceDeleted: "space:deleted",
    SpaceArchived: "space:archived",
    SpaceExpired: "space:expired",
    SpaceGuestAdded: "space:guestAdded",
    SpaceGuestRemoved: "space:guestRemoved",
    DocumentCreated: "document:created",
    DocumentUpdated: "document:updated",
    DocumentDeleted: "document:deleted",
    WorkbookCreated: "workbook:created",
    WorkbookUpdated: "workbook:updated",
    WorkbookDeleted: "workbook:deleted",
    WorkbookExpired: "workbook:expired",
    SheetCreated: "sheet:created",
    SheetUpdated: "sheet:updated",
    SheetDeleted: "sheet:deleted",
    SheetCountsUpdated: "sheet:counts-updated",
    SnapshotCreated: "snapshot:created",
    RecordsCreated: "records:created",
    RecordsUpdated: "records:updated",
    RecordsDeleted: "records:deleted",
    FileCreated: "file:created",
    FileUpdated: "file:updated",
    FileDeleted: "file:deleted",
    FileExpired: "file:expired",
    JobCreated: "job:created",
    JobUpdated: "job:updated",
    JobDeleted: "job:deleted",
    JobCompleted: "job:completed",
    JobReady: "job:ready",
    JobScheduled: "job:scheduled",
    JobOutcomeAcknowledged: "job:outcome-acknowledged",
    JobPartsCompleted: "job:parts-completed",
    JobFailed: "job:failed",
    ProgramCreated: "program:created",
    ProgramUpdated: "program:updated",
    CommitCreated: "commit:created",
    CommitUpdated: "commit:updated",
    CommitCompleted: "commit:completed",
    LayerCreated: "layer:created",
    SecretCreated: "secret:created",
    SecretUpdated: "secret:updated",
    SecretDeleted: "secret:deleted",
    Cron5Minutes: "cron:5-minutes",
    CronHourly: "cron:hourly",
    CronDaily: "cron:daily",
    CronWeekly: "cron:weekly",
    EnvironmentCreated: "environment:created",
    EnvironmentUpdated: "environment:updated",
    EnvironmentDeleted: "environment:deleted",
};
