// @flatfile/api/dist/serialization/resources/mapping/types/Program.js
"use strict";
/**
 * This file was auto-generated by Fern from our API Definition.
 */
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || function (mod) {
    if (mod && mod.__esModule) return mod;
    var result = {};
    if (mod != null) for (var k in mod) if (k !== "default" && Object.prototype.hasOwnProperty.call(mod, k)) __createBinding(result, mod, k);
    __setModuleDefault(result, mod);
    return result;
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.Program = void 0;
const core = __importStar(require("../../../../core"));
const MappingRuleOrConfig_1 = require("./MappingRuleOrConfig");
const FamilyId_1 = require("../../commons/types/FamilyId");
const UserId_1 = require("../../commons/types/UserId");
const ProgramSummary_1 = require("./ProgramSummary");
exports.Program = core.serialization.object({
    rules: core.serialization.list(MappingRuleOrConfig_1.MappingRuleOrConfig),
    id: core.serialization.string().optional(),
    namespace: core.serialization.string().optional(),
    familyId: FamilyId_1.FamilyId.optional(),
    createdAt: core.serialization.date().optional(),
    createdBy: UserId_1.UserId.optional(),
    sourceKeys: core.serialization.list(core.serialization.string()),
    destinationKeys: core.serialization.list(core.serialization.string()),
    summary: ProgramSummary_1.ProgramSummary.optional(),
    accessToken: core.serialization.string().optional(),
});
