// @flatfile/api/dist/serialization/resources/spaces/types/InternalSpaceConfigBase.js
"use strict";
/**
 * This file was auto-generated by Fern from our API Definition.
 */
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || function (mod) {
    if (mod && mod.__esModule) return mod;
    var result = {};
    if (mod != null) for (var k in mod) if (k !== "default" && Object.prototype.hasOwnProperty.call(mod, k)) __createBinding(result, mod, k);
    __setModuleDefault(result, mod);
    return result;
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.InternalSpaceConfigBase = void 0;
const core = __importStar(require("../../../../core"));
const SpaceConfigId_1 = require("../../commons/types/SpaceConfigId");
const EnvironmentId_1 = require("../../commons/types/EnvironmentId");
const WorkbookId_1 = require("../../commons/types/WorkbookId");
const SpaceSettings_1 = require("./SpaceSettings");
const Action_1 = require("../../commons/types/Action");
const SpaceAccess_1 = require("./SpaceAccess");
const AppId_1 = require("../../commons/types/AppId");
exports.InternalSpaceConfigBase = core.serialization.object({
    spaceConfigId: SpaceConfigId_1.SpaceConfigId.optional(),
    environmentId: EnvironmentId_1.EnvironmentId.optional(),
    primaryWorkbookId: WorkbookId_1.WorkbookId.optional(),
    metadata: core.serialization.any().optional(),
    settings: SpaceSettings_1.SpaceSettings.optional(),
    actions: core.serialization.list(Action_1.Action).optional(),
    access: core.serialization.list(SpaceAccess_1.SpaceAccess).optional(),
    autoConfigure: core.serialization.boolean().optional(),
    namespace: core.serialization.string().optional(),
    labels: core.serialization.list(core.serialization.string()).optional(),
    translationsPath: core.serialization.string().optional(),
    languageOverride: core.serialization.string().optional(),
    archivedAt: core.serialization.date().optional(),
    appId: AppId_1.AppId.optional(),
});
