// @flatfile/api/serialization/resources/sheets/types/SheetConfigUpdate.js
"use strict";
/**
 * This file was auto-generated by Fern from our API Definition.
 */
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || function (mod) {
    if (mod && mod.__esModule) return mod;
    var result = {};
    if (mod != null) for (var k in mod) if (k !== "default" && Object.prototype.hasOwnProperty.call(mod, k)) __createBinding(result, mod, k);
    __setModuleDefault(result, mod);
    return result;
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.SheetConfigUpdate = void 0;
const core = __importStar(require("../../../../core"));
const SheetAccess_1 = require("./SheetAccess");
const Property_1 = require("../../property/types/Property");
const Action_1 = require("../../commons/types/Action");
exports.SheetConfigUpdate = core.serialization.object({
    name: core.serialization.string().optional(),
    description: core.serialization.string().optional(),
    slug: core.serialization.string().optional(),
    readonly: core.serialization.boolean().optional(),
    allowAdditionalFields: core.serialization.boolean().optional(),
    mappingConfidenceThreshold: core.serialization.number().optional(),
    access: core.serialization.list(SheetAccess_1.SheetAccess).optional(),
    fields: core.serialization.list(Property_1.Property).optional(),
    actions: core.serialization.list(Action_1.Action).optional(),
});
