// @flatfile/api/dist/serialization/resources/commons/types/Action.js
"use strict";
/**
 * This file was auto-generated by Fern from our API Definition.
 */
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || function (mod) {
    if (mod && mod.__esModule) return mod;
    var result = {};
    if (mod != null) for (var k in mod) if (k !== "default" && Object.prototype.hasOwnProperty.call(mod, k)) __createBinding(result, mod, k);
    __setModuleDefault(result, mod);
    return result;
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.Action = void 0;
const core = __importStar(require("../../../../core"));
const ActionMode_1 = require("./ActionMode");
const ActionMessage_1 = require("./ActionMessage");
const ActionSchedule_1 = require("./ActionSchedule");
const InputForm_1 = require("./InputForm");
const ActionConstraint_1 = require("./ActionConstraint");
exports.Action = core.serialization.object({
    slug: core.serialization.string().optional(),
    operation: core.serialization.string().optional(),
    mode: ActionMode_1.ActionMode.optional(),
    label: core.serialization.string(),
    tooltip: core.serialization.string().optional(),
    messages: core.serialization.list(ActionMessage_1.ActionMessage).optional(),
    type: core.serialization.string().optional(),
    description: core.serialization.string().optional(),
    schedule: ActionSchedule_1.ActionSchedule.optional(),
    primary: core.serialization.boolean().optional(),
    confirm: core.serialization.boolean().optional(),
    icon: core.serialization.string().optional(),
    requireAllValid: core.serialization.boolean().optional(),
    requireSelection: core.serialization.boolean().optional(),
    inputForm: InputForm_1.InputForm.optional(),
    constraints: core.serialization.list(ActionConstraint_1.ActionConstraint).optional(),
});
