// @flatfile/api/dist/serialization/resources/auth/types/ApiKey.js
"use strict";
/**
 * This file was auto-generated by Fern from our API Definition.
 */
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || function (mod) {
    if (mod && mod.__esModule) return mod;
    var result = {};
    if (mod != null) for (var k in mod) if (k !== "default" && Object.prototype.hasOwnProperty.call(mod, k)) __createBinding(result, mod, k);
    __setModuleDefault(result, mod);
    return result;
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.ApiKey = void 0;
const core = __importStar(require("../../../../core"));
const ApiKeyId_1 = require("./ApiKeyId");
const RawKey_1 = require("./RawKey");
const ApiKeyType_1 = require("./ApiKeyType");
const EnvironmentId_1 = require("../../commons/types/EnvironmentId");
const AccountId_1 = require("../../commons/types/AccountId");
const ApiKeyOperation_1 = require("./ApiKeyOperation");
exports.ApiKey = core.serialization.object({
    id: ApiKeyId_1.ApiKeyId,
    rawKey: RawKey_1.RawKey.optional(),
    type: ApiKeyType_1.ApiKeyType,
    environmentId: EnvironmentId_1.EnvironmentId.optional(),
    accountId: AccountId_1.AccountId.optional(),
    operations: core.serialization.list(ApiKeyOperation_1.ApiKeyOperation),
    createdAt: core.serialization.date(),
    updatedAt: core.serialization.date().optional(),
    deletedAt: core.serialization.date().optional(),
    secret: core.serialization.string().optional(),
});
