// @flatfile/api/serialization/resources/property/types/BaseProperty.js
"use strict";
/**
 * This file was auto-generated by Fern from our API Definition.
 */
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || function (mod) {
    if (mod && mod.__esModule) return mod;
    var result = {};
    if (mod != null) for (var k in mod) if (k !== "default" && Object.prototype.hasOwnProperty.call(mod, k)) __createBinding(result, mod, k);
    __setModuleDefault(result, mod);
    return result;
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.BaseProperty = void 0;
const core = __importStar(require("../../../../core"));
const Constraint_1 = require("./Constraint");
const FieldAppearance_1 = require("./FieldAppearance");
const Action_1 = require("../../commons/types/Action");
exports.BaseProperty = core.serialization.object({
    key: core.serialization.string(),
    label: core.serialization.string().optional(),
    description: core.serialization.string().optional(),
    constraints: core.serialization.list(Constraint_1.Constraint).optional(),
    readonly: core.serialization.boolean().optional(),
    appearance: FieldAppearance_1.FieldAppearance.optional(),
    actions: core.serialization.list(Action_1.Action).optional(),
    metadata: core.serialization.any().optional(),
    treatments: core.serialization.list(core.serialization.string()).optional(),
    alternativeNames: core.serialization.list(core.serialization.string()).optional(),
});
