// @flatfile/api/serialization/resources/spaces/types/Space.js
"use strict";
/**
 * This file was auto-generated by Fern from our API Definition.
 */
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || function (mod) {
    if (mod && mod.__esModule) return mod;
    var result = {};
    if (mod != null) for (var k in mod) if (k !== "default" && Object.prototype.hasOwnProperty.call(mod, k)) __createBinding(result, mod, k);
    __setModuleDefault(result, mod);
    return result;
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.Space = void 0;
const core = __importStar(require("../../../../core"));
const SpaceId_1 = require("../../commons/types/SpaceId");
const UserId_1 = require("../../commons/types/UserId");
const SpaceSize_1 = require("./SpaceSize");
const GuestAuthenticationEnum_1 = require("../../environments/types/GuestAuthenticationEnum");
const InternalSpaceConfigBase_1 = require("./InternalSpaceConfigBase");
exports.Space = core.serialization
    .object({
    id: SpaceId_1.SpaceId,
    workbooksCount: core.serialization.number().optional(),
    filesCount: core.serialization.number().optional(),
    createdByUserId: UserId_1.UserId.optional(),
    createdByUserName: core.serialization.string().optional(),
    createdAt: core.serialization.date(),
    updatedAt: core.serialization.date(),
    expiredAt: core.serialization.date().optional(),
    lastActivityAt: core.serialization.date().optional(),
    guestLink: core.serialization.string().optional(),
    name: core.serialization.string(),
    displayOrder: core.serialization.number().optional(),
    accessToken: core.serialization.string().optional(),
    isCollaborative: core.serialization.boolean().optional(),
    size: SpaceSize_1.SpaceSize.optional(),
    upgradedAt: core.serialization.date().optional(),
    guestAuthentication: core.serialization.list(GuestAuthenticationEnum_1.GuestAuthenticationEnum),
})
    .extend(InternalSpaceConfigBase_1.InternalSpaceConfigBase);
